tss-esapi/src/structures/buffers/public/rsa.rs

```diff
@@ -0,0 +1,357 @@
+            }
+
+            if self.is_decryption_key && rsa_scheme.algorithm() != RsaSchemeAlgorithm::Null {
+                error!("Invalid rsa scheme algorithm provided with 'restricted' and 'is_decryption_key' set to true");
+                return Err(Error::local_error(WrapperErrorKind::InconsistentParams));
+            }
+        } else {
+            if self.is_decryption_key
+                && self.is_signing_key
+                && rsa_scheme.algorithm() != RsaSchemeAlgorithm::Null
+            {
+                error!("Invalid rsa scheme algorithm provided with 'restricted' set to false and 'is_decryption_key' and 'is_signing_key' set to true");
+                return Err(Error::local_error(WrapperErrorKind::InconsistentParams));
+            }
+            if self.is_signing_key
+                && rsa_scheme.algorithm() != RsaSchemeAlgorithm::RsaPss
+                && rsa_scheme.algorithm() != RsaSchemeAlgorithm::RsaSsa
+                && rsa_scheme.algorithm() != RsaSchemeAlgorithm::Null
+            {
+                error!("Invalid rsa scheme algorithm provided with 'restricted' set to false and 'is_signing_key' set to true");
+                return Err(Error::local_error(WrapperErrorKind::InconsistentParams));
+            }
+
+            if self.is_decryption_key
+                && rsa_scheme.algorithm() != RsaSchemeAlgorithm::RsaEs
+                && rsa_scheme.algorithm() != RsaSchemeAlgorithm::Oaep
+                && rsa_scheme.algorithm() != RsaSchemeAlgorithm::Null
+            {
+                error!("Invalid rsa scheme algorithm provided with 'restricted' set to false and 'is_decryption_key' set to true");
+                return Err(Error::local_error(WrapperErrorKind::InconsistentParams));
+            }
+        }
+
+        Ok(PublicRsaParameters {
+            symmetric_definition_object,
+            rsa_scheme,
+            key_bits,
+            exponent,
+        })
+    }
+}
+
+/// Strucure used to hold the value of a RSA exponent
+#[derive(Clone, Debug, Copy)]
+pub struct RsaExponent {
+    value: u32,
+}
+
+impl RsaExponent {
+    /// Function for creating a new RsaExponent
+    ///
+    /// # Errors
+    /// Will return an error if the value passed into the function
+    /// is not a valid RSA exponent.
+    pub fn create(value: u32) -> Result<Self> {
+        if !RsaExponent::is_valid(value) {
+            error!(
+                "Received invalid value {} when creating rsa exponent",
+                value,
+            );
+            return Err(Error::WrapperError(WrapperErrorKind::InvalidParam));
+        }
+        Ok(RsaExponent { value })
+    }
+
+    /// Method that returns the value of the rsa exponent.
+    pub const fn value(&self) -> u32 {
+        self.value
+    }
+
+    /// Function for checking if a value is valid rsa exponent.
+    pub const fn is_valid(value: u32) -> bool {
+        // TODO: Verify that `value` is a prime number
```

Comment:
I don't think the exponent has to be prime, just coprime with some other value. I think the commonly used values are primes because those work for any sensible RSA private key.

This also means we can't really check whether the value is valid (unless it is prime, but that's not strictly necessary).

Reply:
Hrrm section 12.2.3.5 in the structures part of the spec says "the public exponent. A prime number greater then 2". Have I interpreted this in the wrong way?

Reply:
Oh, it might be that the TPM puts this restriction, I was thinking of RSA in general. In that case it could make sense. Don't know how we'd do it - maybe there are generic implementations out there, or maybe we can store a list of all primes of at most 32 bits.

Reply:
I added the primal crate as dependency in order to do the prime number checking for us.